base/src/Data/Macaw/Syntax/Parser.hs

```diff
@@ -0,0 +1,269 @@
+
+parseNatural :: SExpr.Parser arch ids Atom
+parseNatural = Natural_ <$> TMCL.decimal
+
+parseInteger :: SExpr.Parser arch ids Atom
+parseInteger = TM.try (Integer_ <$> (TMC.char '+' >> TMCL.decimal))
+               <|> (Integer_ <$> (TMC.char '-' >> TMCL.decimal))
+
+parseAtom :: SExpr.Parser arch ids Atom
+parseAtom = TM.try parseKeywordOrAtom
+            <|> TM.try parseNatural
+            <|> TM.try parseInteger
+
+asApp :: SExpr.Syntax Atom -> SExpr.Parser arch ids (SomeTyped (MC.App (MC.Value arch ids)))
+asApp a =
+  case a of
+    SExpr.L [ SExpr.A (Keyword BVAdd), slhs, srhs ] -> do
+      SomeTyped lrep lhs <- asValue slhs
+      SomeTyped rrep rhs <- asValue srhs
+      case (PC.testEquality lrep rrep, lrep) of
+        (Just PC.Refl, MT.BVTypeRepr nr) -> return (SomeTyped lrep (MC.BVAdd nr lhs rhs))
+
+-- | Parse a single type as a 'MT.TypeRepr'
+--
+-- The type forms are:
+--
+--  * @Bool@
+--  * @(BV n)@
+--  * @(Vec n ty)@
+asTypeRepr :: SExpr.Syntax Atom -> Either SExpr.MacawSyntaxError (Some MT.TypeRepr)
+asTypeRepr at =
+  case at of
+    SExpr.A (Keyword Bool_) -> Right (Some MT.BoolTypeRepr)
+    SExpr.L [SExpr.A (Keyword BV_), SExpr.A (Natural_ w)]
+      | Just (WidthRepr nr) <- asWidthRepr w -> Right (Some (MT.BVTypeRepr nr))
+      | otherwise -> Left SExpr.InvalidZeroWidthBV
+    SExpr.L [SExpr.A (Keyword Vec_), SExpr.A (Natural_ w), mty] ->
+      case PN.mkNatRepr w of
+        Some nr ->
+          -- Note that zero-width vectors are technically allowed
+          case asTypeRepr mty of
+            Right (Some ty) -> Right (Some (MT.VecTypeRepr nr ty))
+            Left _errs -> Left (SExpr.InvalidVectorPayload mty)
+    _ -> Left (SExpr.InvalidType at)
+
+-- | Parse the right-hand side of an assignment
+--
+-- Note that it is important that the EvalApp case is last, as there are many
+-- syntactic forms that we might need to dispatch to.
+asRHS
+  :: forall arch ids
+   . SExpr.Syntax Atom
+  -> SExpr.Parser arch ids (Some (MC.AssignRhs arch (MC.Value arch ids)))
+asRHS a =
+  case a of
+    SExpr.L [ SExpr.A (Keyword Undefined), srep ] -> do
+      Some rep <- parseSExpr srep asTypeRepr
+      return (Some (MC.SetUndefined rep))
+    SExpr.L [ SExpr.A (Keyword ReadMemory), saddr, smemRepr ] -> do
+      SomeTyped addrTp addr <- asValue saddr
+      SomeTyped _rep memRepr <- parseSExpr smemRepr asMemRepr
+
+      memWidth <- SExpr.archMemWidth
+      let addrRepr = MT.BVTypeRepr memWidth
+
+      case PC.testEquality addrTp addrRepr of
+        Just PC.Refl -> return (Some (MC.ReadMem addr memRepr))
+        Nothing -> TM.customFailure (SExpr.InvalidAddressWidth a)
+    _ -> do
+      SomeTyped _tp app <- asApp a
+      return (Some (MC.EvalApp app))
+
+-- | Forms:
+--
+--  * @(comment "msg")@
+--  * @(instruction-start addr decoded-asm-text)@
+--  * @(write-memory addr mem-rep value)@
+--  * @(cond-write-memory cond addr mem-rep value)@
+--  * @(reg := rhs)@
```

Comment:
The Lisper in me is somewhat weirded out by the fact that this is `(reg := rhs)` and not `(:= reg rhs)`, especially since this is (AFAICT) the only operator that is given infix treatment.